# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
# Commit: 7a33c8b73972
# Full Hash: 7a33c8b7397200d89a0d3904669777e60550c445
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2025-06-11 18:38:51
# Regressor Bug: 1968812
# File Overlap Count: 1
# Description:
#   Bug 1968812 - Remove the CallWorker thread and use TaskQueue::Observer instead. r=bwc
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D252686
# ==============================================================================

diff -r a3ad3abc4d02 -r 7a33c8b73972 dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Wed Jun 11 13:37:22 2025 +0000
+++ b/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Wed Jun 11 13:37:23 2025 +0000
@@ -13,7 +13,7 @@
 #include "common/browser_logging/WebRtcLog.h"
 #include "gmp-video-decode.h"  // GMP_API_VIDEO_DECODER
 #include "gmp-video-encode.h"  // GMP_API_VIDEO_ENCODER
-#include "libwebrtcglue/CallWorkerThread.h"
+#include "libwebrtcglue/WebrtcTaskQueueWrapper.h"
 #include "modules/audio_device/include/fake_audio_device.h"
 #include "modules/audio_processing/include/audio_processing.h"
 #include "modules/audio_processing/include/aec_dump.h"
@@ -23,6 +23,7 @@
 #include "mozilla/glean/DomMediaWebrtcMetrics.h"
 #include "mozilla/dom/RTCStatsReportBinding.h"
 #include "nsCRTGlue.h"
+#include "nsIEventTarget.h"
 #include "nsIIOService.h"
 #include "nsIObserver.h"
 #include "nsIObserverService.h"
@@ -494,27 +495,20 @@
     audioStateConfig.audio_device_module =
         new rtc::RefCountedObject<FakeAudioDeviceModule>();
 
-    SharedThreadPoolWebRtcTaskQueueFactory taskQueueFactory;
     constexpr bool supportTailDispatch = true;
-    // Note the NonBlocking DeletionPolicy!
-    // This task queue is passed into libwebrtc as a raw pointer.
+    // This task queue is passed into libwebrtc by means of
+    // webrtc::TaskQueueBase::GetCurrent() while running on it.
     // WebrtcCallWrapper guarantees that it outlives its webrtc::Call instance.
-    // Outside of libwebrtc we must use ref-counting to either the
-    // WebrtcCallWrapper or to the CallWorkerThread to keep it alive.
-    auto callWorkerThread = WrapUnique(
-        taskQueueFactory
-            .CreateWebrtcTaskQueueWrapper<DeletionPolicy::NonBlocking>(
-                "CallWorker", supportTailDispatch,
-                webrtc::TaskQueueFactory::Priority::NORMAL,
-                MediaThreadType::WEBRTC_CALL_THREAD)
-            .release());
+    // Outside of libwebrtc it works as a regular TaskQueue.
+    auto callWorkerThread = CreateWebrtcTaskQueueWrapper(
+        GetMediaThreadPool(MediaThreadType::WEBRTC_CALL_THREAD),
+        "CallWorker"_ns, supportTailDispatch);
 
     UniquePtr<webrtc::FieldTrialsView> trials =
         WrapUnique(new MozTrialsConfig());
 
     mSharedWebrtcState = MakeAndAddRef<SharedWebrtcState>(
-        new CallWorkerThread(std::move(callWorkerThread)),
-        std::move(audioStateConfig),
+        std::move(callWorkerThread), std::move(audioStateConfig),
         already_AddRefed(CreateBuiltinAudioDecoderFactory().release()),
         std::move(trials));
     StartTelemetryTimer();